### ISSUE TITLE
Refactor DAO

### ISSUE THREAD

**Issue by stojanov-igor** [UniversalDot/UniversalDot-node#21]
- [x] Organize the DAO pallet in a better way. 
Create a ** Organization Struct** with these properties: 
<meta charset="utf-8"><b style="font-weight:normal;" id="docs-internal-guid-65b3942c-7fff-3f9b-5c3f-9c0ff503e454"><div dir="ltr" style="margin-left:0pt;" align="left">

Characteristic | Description | Type | Entry
-- | -- | -- | --
ID | Unique identifier for an organization | UUID or similar | Automatic
Name | The name of the organization | String | Manual, Mandatory
Description | Basic description regarding the organization, industry, and goals | String | Manual, Optional
Owner | The account that owns the organization. The initial owner is the founder. Ownership should be able to be transferred to other accounts. | Account ID | Automatic, Mandatory
Vision | Document that describes company Vision | String (Hash to IPFS Document) | Manual, Mandatory
Members | Members that belong to an organization | Array of AccountID | Manual, Mandatory
Tasks | Tasks that belong to a certain Organization | Array of TaskID | Manual, Mandatory
Applicants | Users that have applied to join to a certain organization | Array of AccountID | Automatic
Created | The date when the organization was created | DateTime, Block | Automatic
LastUpdated | The date when the organization had an update | DateTime, Block | Automatic
Properties | Custom collection of properties that can be added. | An array of Objects | Manual, Optional

</div></b>

The list of properties is also available [here](https://docs.google.com/document/d/12zZkoJF6duqeT9yr-mBJIsl4R0sPVExu/edit?usp=sharing&ouid=105283746296807145921&rtpof=true&sd=true)

The Dao can currently perform the following functions: _create_vision,  remove_vision ,sign_vision, unsign_vision ,create_organization ,add_members , remove_members ,dissolve_organization, add_tasks , remove_tasks_ 

- [x] Refactor already existing functions to update Organization Struct Storage. Update tests.

The Dao should be refactored to perform the following functions in addition to the ones above:
- [ ] _update_organization_ - extrinsic that enables the user to update the Organization Struct
- [x] _transfer_ownership_ - extrinsic that allows for ownership of the organization to change hands. The initial owner is the founder. The DID pallet might be helpful here.
- [ ] _suspend_organization_? - The organization can be suspended by a sudo account. This function is useful to have, but in later versions, we will remove the sudo account. 

- [x] add unit tests. The test coverage should be > 90.

Note: You can use the following pallet as inspiration: https://github.com/nusantarachain/nuchain/blob/master/frame/organization/src/lib.rs

Budget: $80.

**Comment by vivekvpandya** [comment]
DAO uses lots of `Vec<>` in storage , in Substrate world it is recommended to use `BoundedVec<>` with a Max possible count set in Runtime.

**Comment by stojanov-igor** [comment]
Yes. I will have to create a proper design for how to refactor the Dao. 

We are currently working with a design team on the front-end USER experience. Once this is finished, I will create a proper design for refactoring this pallet. 

Then you can start with development...

**Comment by vivekvpandya** [comment]
When I try to use pallet-did in dao pallet I am getting following error with simple step to just use pallet-did as dependency and use it in Mock.
```
error: failed to select a version for `parity-util-mem`.
    ... required by package `sp-core v3.0.0`
    ... which satisfies dependency `sp-core = "^3.0.0"` of package `frame-support v3.0.0`
    ... which satisfies dependency `frame-support = "^3.0.0"` of package `frame-system v3.0.0`
    ... which satisfies dependency `frame-system = "^3.0.0"` of package `pallet-did v3.0.0 (https://github.com/substrate-developer-hub/pallet-did#8510c92b)`
    ... which satisfies git dependency `pallet-did` of package `pallet-dao v0.0.25 (/home/vivek/dev/univarsal_dot/universal-dot-node/pallets/dao)`
    ... which satisfies path dependency `pallet-dao` (locked to 0.0.25) of package `node-template-runtime v4.0.0-dev (/home/vivek/dev/univarsal_dot/universal-dot-node/runtime)`
    ... which satisfies path dependency `node-template-runtime` (locked to 4.0.0-dev) of package `node-template v4.0.0-dev (/home/vivek/dev/univarsal_dot/universal-dot-node/node)`
versions that meet the requirements `^0.9.0` are: 0.9.0

the package `parity-util-mem` links to the native library `parity-util-mem-ban-duplicates`, but it conflicts with a previous package which links to `parity-util-mem-ban-duplicates` as well:
package `parity-util-mem v0.10.2`
    ... which satisfies dependency `parity-util-mem = "^0.10.0"` (locked to 0.10.2) of package `sc-informant v0.10.0-dev (https://github.com/paritytech/substrate.git?tag=monthly-2021-11#26d69bcb)`
    ... which satisfies git dependency `sc-informant` (locked to 0.10.0-dev) of package `sc-service v0.10.0-dev (https://github.com/paritytech/substrate.git?tag=monthly-2021-11#26d69bcb)`
    ... which satisfies git dependency `sc-service` (locked to 0.10.0-dev) of package `frame-benchmarking-cli v4.0.0-dev (https://github.com/paritytech/substrate.git?tag=monthly-2021-11#26d69bcb)`
    ... which satisfies git dependency `frame-benchmarking-cli` (locked to 4.0.0-dev) of package `node-template v4.0.0-dev (/home/vivek/dev/univarsal_dot/universal-dot-node/node)`
Only one package in the dependency graph may specify the same links value. This helps ensure that only one copy of a native library is linked in the final binary. Try to adjust your dependencies so that only one package uses the links ='parity-util-mem' value. For more information, see https://doc.rust-lang.org/cargo/reference/resolver.html#links.

failed to select a version for `parity-util-mem` which could resolve this conflict
```
Here if I understand correctly we need to update dependencies of did pallet so that there are no more conflicts.

**Comment by stojanov-igor** [comment]
> Here if I understand correctly we need to update dependencies of did pallet so that there are no more conflicts.

Yes. I expect there to be conflicts in dependencies. 

I have placed this issue on hold for now... Focus on implementing the Task deadline feature.  https://github.com/UniversalDot/universal-dot-node/issues/23

**Comment by vivekvpandya** [comment]
I have updated pallet-did https://github.com/vivekvpandya/pallet-did
and with that I am able to compile pallet-dao while did included. https://github.com/UniversalDot/universal-dot-node/pull/27
I will try for organization ownership soon on above PR.

**Comment by stojanov-igor** [comment]
The organization pallet from nuchain is very similar to what we need for the DAO. I will create a design inspired by  https://github.com/nusantarachain/nuchain/blob/master/frame/organization/src/lib.rs

**Comment by stojanov-igor** [comment]
Hi @vivekvpandya..

Did you start working on this issue?

**Comment by vivekvpandya** [comment]
> Hi @vivekvpandya..
> 
> Did you start working on this issue?

Not yet. I will start mostly in this weekend.

**Comment by vivekvpandya** [comment]
IMO all fields which are Array or Vector, that should be kept out of the Dao structure otherwise it makes read operation for whole Dao structure very heavy. Like even if we only want to query organization name or id etc it has to read everything from DB.

**Comment by vivekvpandya** [comment]
Can a user belong to more than one Dao?

**Comment by stojanov-igor** [comment]
> Can a user belong to more than one Dao?

Yes.
